pytorch_lightning/trainer/trainer.py

```diff
@@ -505,31 +522,56 @@ def fit(
         if self._state != TrainerState.INTERRUPTED:
             self._state = TrainerState.FINISHED
 
-        self._set_wide_running_stage(None)
+        self._set_running_stage(None, model)
 
-        return results or 1
+        return self.training_type_plugin.results or 1
 
-    def _set_wide_running_stage(self, stage):
-        model_ref = self.get_model()
+    def pre_dispatch(self):
+        self.training_type_plugin.pre_dispatch()
+        self.precision_plugin.pre_dispatch()
+
+    def post_dispatch(self):
+        self.training_type_plugin.post_dispatch()
+        self.precision_plugin.post_dispatch()
+        self.accelerator_backend.teardown()
 
-        if stage is None:
-            self._running_stage = stage
-            model_ref.running_stage = stage
-            return
+    def dispatch(self):
+        if self.training or self.evaluating:
+            self.training_type_plugin.start_training(self)
 
-        # todo: clean up this routing mess.
-        if self._running_stage == RunningStage.TESTING:
-            stage = RunningStage.TESTING
+        elif self.testing:
+            self.training_type_plugin.start_testing(self)
 
-        # WARNING: With predicting,
-        # trainer _running_state should be RunningStage.TESTING
-        # however, the model running_stage should be RunningStage.PREDICTING or None
-        if model_ref is not None:
-            if self._predicting:
-                model_ref.running_stage = RunningStage.PREDICTING
-            else:
-                model_ref.running_stage = stage
+        elif self.predicting:
+            self.training_type_plugin.start_predicting(self)
+
+        else:
+            raise MisconfigurationException(
+                f"Received {self._running_stage}. Please, open an issue, as it shouldn't happen."
+            )
```

Comment:
This is odd, can't we ensure with tests that this doesn't happen?

Reply:
Removed !